File: Packages/GlobyCore/Sources/GlobyCore/Clock.swift
import Foundation

/// Tempo iniettabile: backoff e cursori non devono dipendere dall'orologio di sistema nei test.
public protocol Clock: Sendable {
    var now: Date { get }
    func sleep(seconds: TimeInterval) async throws
}

public struct SystemClock: Clock {
    public init() {}

    public var now: Date { Date() }

    public func sleep(seconds: TimeInterval) async throws {
        try await Task.sleep(for: .seconds(seconds))
    }
}

File: desktop/globy-core/src/clock.rs
use chrono::{DateTime, Utc};
use std::time::Duration;

/// Tempo iniettabile: backoff e cursori non devono dipendere dall'orologio di sistema nei test.
pub trait Clock: Send + Sync {
    fn now(&self) -> DateTime<Utc>;
    fn sleep(&self, seconds: f64) -> impl Future<Output = ()> + Send;
}

#[derive(Debug, Default, Clone, Copy)]
pub struct SystemClock;

impl Clock for SystemClock {
    fn now(&self) -> DateTime<Utc> {
        Utc::now()
    }

    fn sleep(&self, seconds: f64) -> impl Future<Output = ()> + Send {
        tokio::time::sleep(Duration::from_secs_f64(seconds.max(0.0)))
    }
}
